cvat/apps/engine/views.py

```diff
@@ -851,16 +859,179 @@ def jobs(self, request, pk):
 
         return Response(serializer.data)
 
+    def _is_data_uploading(self) -> bool:
+        return 'data' in self.action
+
     # UploadMixin method
     def get_upload_dir(self):
         if 'annotations' in self.action:
             return self._object.get_tmp_dirname()
-        elif 'data' in self.action:
+        elif self._is_data_uploading():
             return self._object.data.get_upload_dirname()
         elif 'backup' in self.action:
             return backup.get_backup_dirname()
         return ""
 
+    def _get_tus_input_ordering_metafile_name(self) -> str:
+        return "__tus_input_order__"
+
+    def _get_tus_custom_ordering_metafile_name(self) -> str:
+        return "__tus_custom_order__"
+
+    def _get_tus_metafile_names(self) -> List[str]:
+        return [
+            self._get_tus_input_ordering_metafile_name(),
+            self._get_tus_custom_ordering_metafile_name(),
+        ]
+
+    # UploadMixin method
+    def validate_uploaded_file_name(self, filename: str) -> bool:
+        return (
+            super().validate_uploaded_file_name(filename) and
+            osp.basename(filename) not in self._get_tus_metafile_names()
+        )
+
+    def _read_tus_upload_meta_file(self, path: str) -> Sequence[str]:
+        """
+        Reads the input file and returns its contents.
+        """
+
+        with open(path) as f:
+            return [line.strip() for line in f]
+
+    def _prepare_upload_metafile_entry(self, filename: str) -> str:
+        filename = osp.normpath(filename.strip())
+        upload_dir = self.get_upload_dir()
+        if filename.startswith(upload_dir):
+            filename = osp.relpath(filename, upload_dir)
+        return filename.lstrip("/.")
+
+    def _maybe_append_tus_upload_metafile_entry(self,
+        metafile: str, filename: str, *, existing_files: Optional[Sequence[str]] = None
+    ):
+        if existing_files is None:
+            existing_files = self._read_tus_upload_meta_file(metafile)
+
+        new_entry = self._prepare_upload_metafile_entry(filename)
+        if new_entry not in existing_files:
+            with open(metafile, 'a') as f:
+                print(new_entry, file=f)
+        else:
+            # Don't do anything in the opposite case, because
+            # file uploading can be restarted. In such case we'll
+            # just reuse the first position of this file
+            pass
+
+    class _InvalidMetafileError(Exception):
+        """
+        Indicates an invalid upload metafile found
+        """
+
+    def _sort_uploaded_files(self, uploaded_files: List[Dict[str, Any]]) -> List[str]:
```

Comment:
The type of `uploaded_files` doesn't seem right, given the usage - shouldn't it be `List[str]`?

Reply:
Fixed